File: panopticon_py/contracts.py
from __future__ import annotations

from dataclasses import dataclass, asdict
from typing import Any, Literal, TypedDict
from uuid import uuid4

from panopticon_py.time_utils import utc_now_rfc3339_ms

Layer = Literal["L1", "L2", "L3", "L4", "L5"]


class L1Payload(TypedDict, total=False):
    delta_h: float
    ofi: float
    lambda_impact: float
    best_bid: float
    best_ask: float
    latency_ms: float


class L2Payload(TypedDict, total=False):
    trust_score: float
    signal_reliability: float
    external_event_score: float
    sentiment_score: float
    evidence_vector: dict[str, float]


class L3Payload(TypedDict, total=False):
    prior_probability: float
    likelihood_ratio: float
    posterior_probability: float
    ev_net: float
    kelly_fraction: float
    action: str


@dataclass(frozen=True)
class EventEnvelope:
    event_id: str
    layer: Layer
    event_type: str
    event_ts: str
    ingest_ts_utc: str
    source: str
    version_tag: str
    payload: dict[str, Any]
    source_event_id: str | None = None
    market_id: str | None = None
    asset_id: str | None = None

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)


def utc_now_iso() -> str:
    return utc_now_rfc3339_ms()


def build_event(
    *,
    layer: Layer,
    event_type: str,
    source: str,
    version_tag: str,
    payload: dict[str, Any],
    source_event_id: str | None = None,
    market_id: str | None = None,
    asset_id: str | None = None,
) -> EventEnvelope:
    now = utc_now_iso()
    return EventEnvelope(
        event_id=str(uuid4()),
        layer=layer,
        event_type=event_type,
        event_ts=now,
        ingest_ts_utc=now,
        source=source,
        source_event_id=source_event_id,
        version_tag=version_tag,
        market_id=market_id,
        asset_id=asset_id,
        payload=payload,
    )

File: panopticon_ts/src/contracts.ts
export type Layer = "L1" | "L2" | "L3" | "L4" | "L5";

export interface EventEnvelope<TPayload> {
  event_id: string;
  layer: Layer;
  event_type: string;
  event_ts: string;
  ingest_ts_utc: string;
  source: string;
  source_event_id?: string | null;
  version_tag: string;
  market_id?: string | null;
  asset_id?: string | null;
  payload: TPayload;
}

export interface L1Payload {
  delta_h: number;
  ofi: number;
  lambda_impact: number;
  best_bid: number;
  best_ask: number;
  latency_ms: number;
}

export interface L4Payload {
  decision_id: string;
  accepted: boolean;
  reason: string;
  simulated_fill_price?: number;
  simulated_fill_size?: number;
  impact_pct?: number;
}

export function buildEvent<TPayload>(
  layer: Layer,
  eventType: string,
  source: string,
  versionTag: string,
  payload: TPayload,
): EventEnvelope<TPayload> {
  const ts = new Date().toISOString();
  const randomId = `${Date.now()}-${Math.floor(Math.random() * 1_000_000)}`;
  return {
    event_id: randomId,
    layer,
    event_type: eventType,
    event_ts: ts,
    ingest_ts_utc: ts,
    source,
    version_tag: versionTag,
    payload,
  };
}
